crates/matrix-sdk/src/authentication/oauth/qrcode/reciprocate/scanned.rs

```diff
@@ -0,0 +1,728 @@
+}
+
+impl std::fmt::Debug for ReciprocateHandler {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReciprocateHandler")
+            .field("secrets_bundle", &"SecretsBundle")
+            .field("channel", &"Channel")
+            .field("qr_code_data", &self.qr_code_data)
+            .field("state", &self.state)
+            .finish()
+    }
+}
+
+impl ReciprocateHandler {
+    /// Subscribe to the progress of the QR code login.
+    pub fn subscribe_to_progress(&self) -> impl Stream<Item = ReciprocateProgress> + use<> {
+        self.state.subscribe()
+    }
+
+    /// Create a new handler.
+    ///
+    /// This will create a rendezvous session and an associated QR code that can
+    /// be scanned by the new device.
+    pub(crate) async fn new(client: Client) -> Result<Self, Error> {
+        // Create a new ephemeral key pair and a rendezvous session to reciprocate the
+        // login with.
+        // -- MSC4108 Secure channel setup steps 1 & 2
+        let homeserver_url = client.homeserver();
+        let http_client = client.inner.http_client.clone();
+        let secure_channel = SecureChannel::reciprocate(http_client, &homeserver_url).await?;
+
+        // Extract the QR code data so that the caller can present the QR code for
+        // scanning by the new device. -- MSC4108 Secure channel setup step 3
+        let qr_code_data = secure_channel.qr_code_data().clone();
+
+        let channel = Channel::Insecure(secure_channel);
+        let secrets_bundle = export_secrets_bundle(&client).await?;
+
+        Ok(Self { client, channel, qr_code_data, secrets_bundle, state: Default::default() })
+    }
+
+    /// Get the data for the QR code that the new device needs to scan.
+    pub fn qr_code_data(&self) -> &QrCodeData {
+        &self.qr_code_data
+    }
+
+    /// Wait for the new device to scan the QR code.
+    pub async fn wait_for_scan(&mut self) -> Result<(), Error> {
+        let Channel::Insecure(channel) = self.channel.take() else {
+            return Err(Error::InvalidState);
+        };
+
+        // Wait for the secure channel to connect. The other device now needs to scan
+        // the QR code and send us the LoginInitiateMessage which we respond to
+        // with the LoginOkMessage. -- MSC4108 step 4 & 5
+        self.channel = Channel::Almost(channel.connect().await?);
+
+        // The other device now needs to verify our message, compute the checkcode and
+        // display it. We emit a progress update to let the caller prompt the
+        // user to enter the checkcode and feed it back to us.
+        // -- MSC4108 Secure channel setup step 6
+        self.state.set(ReciprocateProgress::WaitingForCheckCode);
+
+        Ok(())
+    }
+
+    /// Confirm the check code entered by the user and finish the reciprocation.
+    pub async fn confirm_check_code_and_finish(&mut self, check_code: u8) -> Result<(), Error> {
```

Comment:
Is there a reason for this API to be inconsistent with your previous PR?

Why do we have two methods that need to be called instead of using a check-code sender again?

Having two steps here means that you need to account for the channel being used up, or `Poisoned`, it also means that you need to clone the QR code data so you have always access to it.

If we reuse the patterns we have for the QR login we would have a named future that allows you to get the QR code data and subscribe and once you await all this data gets moved into the future and you don't need to have the `Poisoned` state nor the clone of the QR code data.

Finally the `SecretsBundle` could be moved into `finish_login_grant` instead of passing a borrow, avoiding another clone.

Reply:
Hm, I think you're making a convincing case for why the other / existing API shape is better. Let me try to swap this around.